llvm/tools/sycl-split/sycl-split.cpp

```diff
@@ -0,0 +1,263 @@
+  OS.close();
+}
+
+static void collectKernelsSet(
+    Module &M,
+    std::map<std::string, std::vector<Function *>> &ResKernelsSet) {
+  for (auto &F : M.functions()) {
+    if (F.getCallingConv() != CallingConv::SPIR_KERNEL)
+      continue;
+
+    if (F.hasFnAttribute("module-id")) {
+      Attribute Id = F.getFnAttribute("module-id");
+      std::string Val = Id.getValueAsString();
+      ResKernelsSet[Val].push_back(&F);
+    }
+  }
+}
+
+static void
+splitModule(Module &M,
+            std::map<std::string, std::vector<Function *>> &KernelsSet,
+            std::vector<std::unique_ptr<Module>> &ResModules,
+            std::vector<std::string> &ResSymbolsLists) {
+  for (auto &It : KernelsSet) {
+    // For each group of kernels collect all dependencies.
+    SetVector<GlobalValue *> GVs;
+    std::vector<llvm::Function *> Workqueue;
+    std::string SymbolsList;
+
+    for (auto &F : It.second) {
+      GVs.insert(F);
+      Workqueue.push_back(F);
+      SymbolsList =
+          (Twine(SymbolsList) + Twine(F->getName()) + Twine("\n")).str();
+    }
+
+    while (!Workqueue.empty()) {
+      Function *F = &*Workqueue.back();
+      Workqueue.pop_back();
+      for (auto &BB : *F) {
+        for (auto &I : BB) {
+          CallBase *CB = dyn_cast<CallBase>(&I);
+          if (!CB)
+            continue;
+          Function *CF = CB->getCalledFunction();
+          if (!CF)
+            continue;
+          if (CF->isDeclaration() || GVs.count(CF))
+            continue;
+          GVs.insert(CF);
+          Workqueue.push_back(CF);
+        }
+      }
+    }
+
+    // It's not easy to trace global variable's uses inside needed functions
+    // because global variable can be used inside a combination of operators, so
+    // mark all global variables as needed and remove dead ones after
+    // extraction.
+    for (auto &G : M.globals()) {
+      GVs.insert(&G);
+    }
+
+    // Clone the module, understand which globals we need to extract from the
+    // clone.
+    ValueToValueMapTy VMap;
+    std::unique_ptr<Module> MClone = CloneModule(M, VMap);
+    std::vector<GlobalValue *> GVsInClone(GVs.size());
+    int I = 0;
+    for (GlobalValue *GV : GVs) {
+      GVsInClone[I] = cast<GlobalValue>(VMap[GV]);
+      ++I;
+    }
+
+    // TODO: Use the new PassManager instead?
+    legacy::PassManager Extract;
+
+    // Extract needed globals.
+    Extract.add(createGVExtractionPass(GVsInClone, /* deleteS */ false));
+    Extract.run(*MClone.get());
+
+    // Extactor pass sets external linkage to all globals. Return linkage back.
+    for (auto &G : MClone->globals()) {
+      if (G.getVisibility() == GlobalValue::HiddenVisibility) {
+        G.setLinkage(GlobalValue::InternalLinkage);
+      }
+    }
+
+    legacy::PassManager Passes;
+    // Do cleanup.
+    Passes.add(createGlobalDCEPass());           // Delete unreachable globals.
+    Passes.add(createStripDeadDebugInfoPass());  // Remove dead debug info.
```

Comment:
Ah, this is probably the cause of a long-standing bug in the similar approach we have in triSYCL https://github.com/triSYCL/llvm/blob/sycl/master/lib/SYCL/SYCLKernelFilter.cpp !
It crashes when compiling with `-g`... :-(
Interesting to know... :-)